psci/Main.hs

```diff
@@ -0,0 +1,349 @@
+            P.parseForeignModulesFromFiles foreignFilesContent
+          (externs, env) <- ExceptT . runMake . make (initialPSCiState { psciForeignFiles = foreigns }) . map snd $ allModules
+          return (allModules, foreigns, externs, env)
+        case e of
+          Left errs -> putStrLn (P.prettyPrintMultipleErrors False errs) >> exitFailure
+          Right (modules, foreigns, externs, env) -> do
+            historyFilename <- getHistoryFilename
+            let settings = defaultSettings { historyFile = Just historyFilename }
+            flip evalStateT (PSCiState [] inputFiles (zip (map snd modules) externs) foreigns [] psciInputNodeFlags env) . runInputT (setComplete completion settings) $ do
+              outputStrLn prologueMessage
+              unless (consoleIsDefined externs) . outputStrLn $ unlines
+                [ "PSCi requires the purescript-console module to be installed."
+                , "For help getting started, visit http://wiki.purescript.org/PSCi"
+                ]
+              go
+      where
+        go :: InputT (StateT PSCiState IO) ()
+        go = do
+          c <- getCommand (not psciMultiLineMode)
+          case c of
+            Left err -> outputStrLn err >> go
+            Right Nothing -> go
+            Right (Just QuitPSCi) -> outputStrLn quitMessage
+            Right (Just c') -> do
+              handleInterrupt (outputStrLn "Interrupted.")
+                              (withInterrupt (runPSCI (handleCommand c')))
+              go
+
+-- Compile the module
+
+-- |
+-- Load all modules, updating the application state
+--
+loadAllImportedModules :: PSCI ()
+loadAllImportedModules = do
+  files <- gets psciLoadedFiles
+  e <- runExceptT $ do
+    modules <- ExceptT . liftIO $ loadAllModules files
+    st <- get
+    (externs, _) <- ExceptT . liftIO . runMake . make st . map snd $ modules
+    return (map snd modules, externs)
+  case e of
+    Left errs -> PSCI $ printErrors errs
+    Right (modules, externs) -> modify . updateLoadedExterns $ zip modules externs
+
+-- | This is different than the runMake in 'Language.PureScript.Make' in that it specifies the
+-- options and ignores the warning messages.
+runMake :: P.Make a -> IO (Either P.MultipleErrors a)
+runMake mk = fst <$> P.runMake P.defaultOptions mk
+
+-- | Read a file in the 'P.Make' monad, handling errors.
+readFileMake :: FilePath -> P.Make String
+readFileMake path = P.makeIO (const (P.ErrorMessage [] $ P.CannotReadFile path)) (readUTF8File path)
+
+-- | Rebuild a module, using the cached externs data for dependencies.
+rebuild :: PSCiState -> P.Module -> P.Make (P.ExternsFile, P.Environment)
+rebuild PSCiState{..} m = do
+    let loadedExterns = map snd psciLoadedExterns
+    externs <- P.rebuildModule buildActions loadedExterns m
+    return (externs, foldl' (flip P.applyExternsFileToEnvironment) P.initEnvironment (loadedExterns ++ [externs]))
```

Comment:
I like the second option. I'd like to avoid passing around environments though, since those are really an internal structure for the typechecker. Externs should capture everything in the public API for a module, so if we accidentally end up with more stuff in the `Environment` after typechecking, it's fine, because we'll reconstruct it from the externs anyway.